### PR TITLE
API Introduce thumbnail loading mechanism

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/silverstripe-framework/issues/5544

In-progress images look like this (animated)

image

Images that failed after the specified retry period look like this

image

@clarkepaul can you please review the styles of the screenshots I posted?

### CODE REVIEW

**Comment by clarkepaul** [comment]
Is it too late to use the button loading in place of the S loading? I've had feedback that it sends bad connotations having the loading S all over the place (ect. they think our system is slow rather than their hosting). So I'd be trying not to use the S for new scenarios but rather only use it in the context of loading panels as we currently do.

File: client/src/state/imageLoad/ImageLoadActionHandler.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  expiry: 0,
+};
+
+class ImageLoadActionHandler {
+
+  constructor(options, action) {
+    this.onStatusChange = null;
+    this.onRetry = null;
+    this.onReset = null;
+    this.onTimeout = null;
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActionHandler.js]: I had set these event callbacks to the `this.options` object, so they can be set by passing in options.
This makes the setters less relevant but still available if needed.

File: client/src/state/imageLoad/ImageLoadActionHandler.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * Change the status of a file
+   *
+   * @param {String} url
+   * @param {String} status
+   */
+  handleStatusChange(url, status) {
+    if (this.onStatusChange) {
+      this.onStatusChange(url, status);
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActionHandler.js]: had deleted `handleStatusChange()` as it was just a pass-through function

File: client/src/state/imageLoad/ImageLoadActionHandler.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    // Otherwise, stop further attempts
+    return null;
+  }
+
+  /**
+   * On failure we want to temporarily mark as failed, and retry later
+   *
+   * @param {String} url
+   * @return {Promise}|{null}
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActionHandler.js]: valid JSDocs is `{Promise|null}`

File: client/src/state/imageLoad/ImageLoadActionHandler.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+   * @param {String} url
+   * @param {String} status
+   */
+  handleStatusChange(url, status) {
+    if (this.onStatusChange) {
+      this.onStatusChange(url, status);
+    }
+  }
+
+  handleTimeout(callback, delay) {
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActionHandler.js]: missing docblock

File: client/src/state/imageLoad/ImageLoadActions.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import ImageLoadActionHandler from './ImageLoadActionHandler';
+
+/**
+ * Default factory to test for image
+ *
+ * @param {String} url
+ * @param {function} resolve
+ * @param {function} reject
+ */
+const defaultImageFactory = (url, resolve, reject) => {
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActions.js]: moved this to `ImageLoadActionHandler` as a default for constructor

File: client/src/state/imageLoad/ImageLoadActions.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      return null;
+    }
+
+    // Only process new urls, or urls that have been reset
+    const state = getState();
+    const currentFile = state
+      && state.assetAdmin
+      && state.assetAdmin.imageLoad
+      && state.assetAdmin.imageLoad.files
+      && state.assetAdmin.imageLoad.files.find((file) => file.url === url);
```
Comment (flamerohr) [inline client/src/state/imageLoad/ImageLoadActions.js]: don't need all those prop.exists checks, as Reducer should've populated all those before action is called.

File: client/src/state/imageLoad/tests/ImageLoadActions-test.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+jest.unmock('react');
+jest.unmock('deep-freeze-strict');
+jest.unmock('../ImageLoadActionHandler');
+jest.unmock('../ImageLoadActionTypes');
+
+import ImageLoadActionHandler from '../ImageLoadActionHandler';
+import IMAGE_STATUS from '../ImageLoadStatus';
+
+describe('imageLoadActions', () => {
```
Comment (flamerohr) [inline client/src/state/imageLoad/tests/ImageLoadActions-test.js]: this is more accurately describing `ImageLoadActionHandler` so better to give that description here

File: client/src/state/imageLoad/tests/ImageLoadActions-test.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
+/* global jest, describe, it, expect, beforeEach, jasmine */
```
Comment (flamerohr) [inline client/src/state/imageLoad/tests/ImageLoadActions-test.js]: Might be worth having a look through the `ImageLoadActionHandler-test.js` file altogether, there were a few things changed which I didn't list out here

**Comment by flamerohr** [comment]
I'd like to add some unit tests for `loadImage` in `ImageLoadActions` before merging :)

**Comment by tractorcow** [comment]
Rebased and requested tests added.

**Pull request merged**